### ISSUE TITLE
[FSDP] Gradients not propagating for mixed precision case

### ISSUE THREAD

**Issue by awgu** [pytorch/pytorch#94857]
```
import torch
from torch import nn
import torch.distributed as dist
torch.use_deterministic_algorithms(True, warn_only=True)
torch.manual_seed(0)
dist.init_process_group(backend="nccl")
global_rank = dist.get_rank()
torch.cuda.set_device(global_rank)
import functools

from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
from torch.distributed.fsdp import ShardingStrategy
from torch.distributed.fsdp import MixedPrecision
from torch.distributed.fsdp.wrap import ModuleWrapPolicy
mixed_precision_config = MixedPrecision(
    param_dtype=torch.float16,
    reduce_dtype=torch.float32,
    buffer_dtype=torch.float32,
)
sharding_strategy_config = ShardingStrategy.SHARD_GRAD_OP
teacher_backbone = nn.Linear(24, 24)
student_backbone = nn.Linear(24, 24)
head = nn.Linear(24, 24)
x_half = torch.randn(32, 24, device="cuda", dtype=torch.float16)
x_half2 = torch.randn(32, 24, device="cuda", dtype=torch.float16)
fsdp_wrapper = functools.partial(
    FSDP, 
    sharding_strategy=sharding_strategy_config,
    mixed_precision=mixed_precision_config,
    device_id=global_rank,
    use_orig_params=True,
)
s_teacher_backbone = fsdp_wrapper(teacher_backbone)
s_student_backbone = fsdp_wrapper(student_backbone)
s_head = fsdp_wrapper(head)
t1 = s_teacher_backbone(x_half)
t2 = s_head(t1).detach()
s1 = s_student_backbone(x_half2)
s2 = s_head(s1)
loss = (s2 * t2).sum()
loss.backward()
```

Even if `t2` has `detach()` called, the gradient should propagate through `s2` to `s_head`. However, none of `s_head`'s parameters receive gradients.

If you remove the FSDP from this repro, the gradients are propagated as expected.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501

**Comment by awgu** [comment]
This is a bug with `use_orig_params=True` + mixed precision.

**Comment by rohan-varma** [comment]
@awgu did we try multi-grad? if that's not the direction we want to go, what might be a fix here?

also, can you explain why this only occurs for parameter mixed precision and not in fp32?

**Comment by awgu** [comment]
@rohan-varma Changing a tensor's data to different precisions changes its `AccumulateGrad` object, which only affects the mixed precision case.

This is both a niche and challenging issue. I am not sure when we will have bandwidth to explore a solution in more depth.

**Comment by eliphatfs** [comment]
I meet this issue too today, when training with a multi-iteration algorithm with only the last iteration non-detached.

**Comment by eliphatfs** [comment]
What are the downsides of always re-registering? What cases will that break?
